### PR TITLE
(RGUI) Performance optimisations

### PR DESCRIPTION
**Pull request by jdgleaver**
## Description

This PR heavily optimises the software rendering aspect of RGUI. It makes the following changes:

- `blit_line()`: When printing text, all the heavy calculations have been replaced with a lookup table

- `blit_line()`: When printing text, we no longer waste CPU cycles 'drawing' whitespace

- `blit_line()`: A little pointer trick has been used to increase the efficiency of drawing text shadows

- Regular (non-wallpaper) backgrounds are now handled in the same way as wallpaper backgrounds - i.e. they are generated/cached only when they change, and are written to the framebuffer using `memcpy()`. This is almost 3x faster than the old drawing method.

- Fullscreen and inline playlist thumbnails are now written line-by-line using `memcpy()`, rather than pixel-by-pixel

- `rgui_render()`: The clock display is now generated directly using `time()` and `localtime()`, rather than the (surprisingly) expensive `menu_display_timedate()` function

- `rgui_render()`: A number of redundant/repeated function calls have been removed

This was profiled with Callgrind under Linux using the `gl` driver with Nvidia hardware. The results are as follows:

- When generating 'normal' (text-only) menus, the performance overheads of `rgui_render()` are reduced by ~74%

- The performance overhead of drawing fullscreen thumbnails is reduced by ~77%

- The performance overhead of drawing inline playlist thumbnails is reduced by ~31%

- Enabling menu shadows only increases performance overheads by ~4%

Note that the actual 'real world' performance gains are harder to quantify, since 'rendering' is only half the story - we still have to put the generated framebuffer on-screen, and this is entirely dependent upon GFX drivers.

For example, on my system with these optimisations in place, almost 95% of the total CPU time is spent inside `gl2_set_texture_frame()` - so RGUI itself is basically having no performance impact at all now. GFX driver stuff is outside of my comfort zone, so I can't really help with optimising this side of things...

---

**SIDE NOTE:**

While working on this, I noticed that RGUI is underutilising its font. What I mean here is that the internal bitmap font supports extended ASCII characters, but we have no way of actually displaying them due to the way that char arrays are formatted. This is an encoding issue that I could (with some effort) handle inside RGUI - but the performance impact would make this untenable.

Maybe this could be handled at a higher level, but it would be quite far reaching...

In the meantime, I have just edited RGUI such that non-standard ASCII characters (with a value above 127) simply are not drawn. This prevents the display of 'garbage' characters, and - more importantly -  means that we can never overflow the framebuffer and get a segfault (which was a very real possibility before).

### CODE REVIEW

**Comment by orbea** [comment]
I'm just curious, are there any real performance differences in context of this PR with `glcore` or `vulkan` compared to the `gl` driver?

**Comment by jdgleaver** [comment]
@orbea With this PR, the performance impact of the RGUI-specific rendering stuff becomes negligible, so almost everything comes down to the act of pushing pixels onto the screen (i.e. just the basic infrastructure stuff that RetroArch always has to do). This is dependent upon (a) the driver selected in RetroArch and (b) the GFX hardware in your system.

We've been looking at this on Discord, and there is indeed a performance difference between the RetroArch drivers. We've been looking at static menu rendering, and on my system with `gl` as a baseline the *increase* in performance overheads is as follows:

- `vulkan`: +7%

- `gl1`: +200%

- `glcore`: +12%

- `sdl2`: -46%

...so `sdl2` is best, `gl1` is worst, and `gl` is the best/most practical choice in most cases.

In terms of real world performance, almost every platform won't even notice that it's running RGUI - the CPU usage is tiny. But on really old, really slow hardware (with very slow RAM), you may notice a difference.

TLDR version:

- RGUI rendering is now basically 'free'

- All performance issues are down to GFX drivers

- You won't notice a difference between GFX drivers unless your hardware is junk

**Comment by orbea** [comment]
Thanks for the explanation, I just want to add one thing.

> gl is the best/most practical choice in most cases.

With the `gl` driver you will be getting a GL compatibility profile while with `glcore` you will be getting the preferably core profile. This is not so much of an issue with newer mesa versions, but before mesa users would be getting a `3.3` core profile when they had a `4.5` core profile available...

> We've been looking at this on Discord

Too bad it couldn't be somewhere more accessible.

**Comment by jdgleaver** [comment]
@orbea:

> With the gl driver you will be getting a GL compatibility profile while with glcore you will be getting the preferably core profile. This is not so much of an issue with newer mesa versions, but before mesa users would be getting a 3.3 compat profile when they had a 4.5 core profile available...

You interpreted my statement too generally...  :slightly_smiling_face: 

I meant that `gl` was 'best' only in the context of running RGUI (which makes heavy use of cached frame drawing when the menu is idle, which has extra overheads in `glcore`). Of course you are correct - in other use cases, other drivers all have their advantages.

> Too bad it couldn't be somewhere more accessible.

I hear you - I don't like any social media stuff, but they asked me to join, so what's a person to do? :slightly_smiling_face: 

There was certainly no intention to conceal information - this was just a discussion that arose after the PR was made. But still - you are an incredibly valuable member of the community, and it's not right or fair that you should feel excluded. Is there any reason you can't use Discord? Can I do anything to help?

**Comment by orbea** [comment]
Understood about in the context of rgui. :)

> Is there any reason you can't use Discord?

I'd rather not use proprietary malware which collects user information in the same way something like facebook does among other things. Also this is not the place for it, but there is rather distasteful information about some of at least some of the discord staff now which makes them even less trustworthy...

> Can I do anything to help?

Its not your fault, I'm just find the situation unfortunate, we used to have a lively community on irc...

**Comment by jdgleaver** [comment]
@orbea I understand you 100%. Such a shame that the internet today has to be tainted like this... :slightly_frowning_face: 

Oh well, we still have github :slightly_smiling_face:

**Comment by twinaphex** [comment]
The only thing I can promise is that I hang around on the IRC more in the future. It is ultimately inconvenient to need a separate app open for it at all times, but I will still try to be there for the purpose of talking to those who don't want anything to do with Discord.
